nav2_smac_planner/src/node_hybrid.cpp

```diff
@@ -492,6 +515,126 @@ float NodeHybrid::getObstacleHeuristic(
   return 2.0 * obstacle_heuristic_lookup_table[start_index];
 }
 
+inline float distanceHeuristic2D(
+  const int idx, const int size_x, const int goal_x,
+  const int goal_y)
+{
+  int x = std::abs(idx % size_x - goal_x);
+  int y = std::abs(idx / size_x - goal_y);
+  float & table_value = NodeHybrid::astar_2d_h_table[y * size_x + x];
+  return table_value < 0.0f ? table_value = std::hypotf(x, y) : table_value;
+}
+
+float NodeHybrid::getObstacleHeuristicAdmissible(
+  const Coordinates & node_coords,
+  const Coordinates & goal_coords,
+  const double & cost_penalty)
+{
+  int size_x = sampled_costmap->getSizeInCellsX();
+  // Divided by 2 due to downsampled costmap.
+  const int start_y = floor(node_coords.y / 2.0);
+  const int start_x = floor(node_coords.x / 2.0);
+  const int start_index = start_y * size_x + start_x;
+  const float & starting_cost = obstacle_heuristic_lookup_table[start_index];
+  if (starting_cost > 0.0f) {
+    // costs are doubled due to downsampling
+    return 2.0 * starting_cost;
+  }
+
+  const int size_x_int = static_cast<int>(size_x);
+  const int size_y = sampled_costmap->getSizeInCellsY();
+
+  const std::vector<int> neighborhood = {1, -1,  // left right
+    size_x_int, -size_x_int,  // up down
+    size_x_int + 1, size_x_int - 1,  // upper diagonals
+    -size_x_int + 1, -size_x_int - 1};  // lower diagonals
+
+  const int goal_x = floor(goal_coords.x / 2.0);
+  const int goal_y = floor(goal_coords.y / 2.0);
+  const int goal_index = goal_y * size_x + goal_x;
+  if (astar_2d_first_run) {
+    astar_2d_first_run = false;
+
+    astar_2d_queue.emplace_back(
+      distanceHeuristic2D(
+        goal_index, size_x, start_x,
+        start_y), goal_index);
+
+    // initialize goal cell with a very small value to differentiate it from 0.0 (~uninitialized)
+    // the negative value means the cell is in the open set
+    obstacle_heuristic_lookup_table[goal_index] = -0.00001f;
+  } else {
+    for (auto & n : astar_2d_queue) {
```

Comment:
Marching through every cell here is costly. I'm not understanding this loop. If we've already started the table and we're still executing the same planning / goal run, why is this being updated? 

The goal pose (e.g. where we start) is still the same, so for any given cell, the heuristic value would continue to be the same. We're continuing to expand the heuristic front because we were asked to give an H value for a cell that wasn't previously assigned in the last expansion. All of the prior expanded cells should still be valid since its the same planning request. Right?

Reply:
I suspect this is not actually required

Reply:
Yes, this is the most questionable element in the new implementation so I'm going to elaborate a bit:

First thing - the heuristic algorithm finds path costs in the opposite direction, i.e. from the goal pose to node. So the goal is the start and current node is the target. All of the prior expanded cell costs are valid since its the same planning request but the 2D A* cell priority is composed of 

**travel cost from start** (start is the goal position so this cost remains the same) + **euclidean distance heuristic towards target** (target is the currently requested node so this **changes**)

The second term causes, unfortunately, that reprioritization is needed. I've done quite a wide research, searching for an algorithm that addresses this problem (path search from an unchanging position to a changing position) better and found multitude of algorithms which address similar problems (moving target replanning/pursuit path search) but every seemingly usable one of them assumed the target doesn't move too fast, which is not our case because our targets (~current requested nodes) can jump all around the OPEN set during the main A* search and such big jumps would significantly impact the computational speed of the algorithms (they would add the complexity of around `O(q*log(q)*r)` where `q` is the queue size and `r` is number of requested unvisited cells). Queue reprioritization, used in the currently proposed algorithm, has a [linear complexity](https://en.cppreference.com/w/cpp/algorithm/make_heap) and thus adds just `O(q*r)`.

I originally thought about making a hybrid search which uses one of the moving target algorithms for small jumps of the target cell (currently requested node) and the queue reprioritization for larger jumps and yes, I think it is possible, e.g. by combining a lazy requeuing method (described in D* Lite paper) with the currently used queue reprioritization (described in Differential A* paper). However, after implementing the reprioritization part, I did some benchmarking which showed that it already works very well and fast in practice, occupying just about 20% of the whole heuristic computation so I decided not to implement the lazy requeuing.

However, situations which need the lazy requeuing to be solved in real-time can theoretically occur but since the update is a little bit heavier, I would wait with it until (if at all) they actually appear in practice.

Reply:
I see, the need to re-prioritize makes sense after I very carefully went through it. A comment at this block of code about that would be nice since that is also subtle.